ledger/simulation/resources.go

```diff
@@ -0,0 +1,395 @@
+	}
+	return a.localTxnResources[txnIndex].addAsset(aid)
+}
+
+func (a *GroupResourceAssignment) hasApp(aid basics.AppIndex, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.hasApp(aid)
+	}
+	return a.localTxnResources[txnIndex].hasApp(aid)
+}
+
+func (a *GroupResourceAssignment) addApp(aid basics.AppIndex, ep *logic.EvalParams, programVersion uint64, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.addApp(aid, ep, programVersion)
+	}
+	return a.localTxnResources[txnIndex].addApp(aid, ep, programVersion)
+}
+
+func (a *GroupResourceAssignment) hasBox(app basics.AppIndex, name string) bool {
+	// all boxes are global, never consult PerTxnResources
+	return a.Resources.hasBox(app, name)
+}
+
+func (a *GroupResourceAssignment) addBox(app basics.AppIndex, name string) bool {
+	// all boxes are global, never consult PerTxnResources
+	return a.Resources.addBox(app, name)
+}
+
+func (a *GroupResourceAssignment) hasHolding(addr basics.Address, aid basics.AssetIndex) bool {
+	// nil map lookup is ok
+	_, ok := a.AssetHoldings[ledgercore.AccountAsset{Address: addr, Asset: aid}]
+	return ok
+}
+
+func (a *GroupResourceAssignment) addHolding(addr basics.Address, aid basics.AssetIndex) bool {
+	// TODO: limit cross-product usage
+	if a.AssetHoldings == nil {
+		a.AssetHoldings = make(map[ledgercore.AccountAsset]struct{})
+	}
+	a.AssetHoldings[ledgercore.AccountAsset{Address: addr, Asset: aid}] = struct{}{}
+	return true
+}
+
+func (a *GroupResourceAssignment) hasLocal(addr basics.Address, aid basics.AppIndex, ep *logic.EvalParams) bool {
+	if ep.GetApplicationAddress(aid) == addr {
+		// The app local of an app and its own account is always available, so don't bother recording it.
+		return true
+	}
+	// nil map lookup is ok
+	_, ok := a.AppLocals[ledgercore.AccountApp{Address: addr, App: aid}]
+	return ok
+}
+
+func (a *GroupResourceAssignment) addLocal(addr basics.Address, aid basics.AppIndex) bool {
+	// TODO: limit cross-product usage
+	if a.AppLocals == nil {
+		a.AppLocals = make(map[ledgercore.AccountApp]struct{})
+	}
+	a.AppLocals[ledgercore.AccountApp{Address: addr, App: aid}] = struct{}{}
+	return true
+}
+
+func (a *GroupResourceAssignment) boxIOBudget(bytesPerBoxRef uint64) uint64 {
+	return uint64(a.startingBoxes+a.Resources.maxPossibleBoxes()) * bytesPerBoxRef
+}
+
+type resourcePolicy struct {
+	assignment                  GroupResourceAssignment
+	ep                          *logic.EvalParams
+	initialBoxSurplusReadBudget *uint64
+	unnamedBoxUsedReadBudget    uint64
+
+	txnRootIndex   int
+	programVersion uint64
+	globalSharing  bool
+}
+
+func newResourcePolicy(ep *logic.EvalParams, groupResult *TxnGroupResult) *resourcePolicy {
+	txns := make([]*transactions.Transaction, len(ep.TxnGroup))
+	for i := range ep.TxnGroup {
+		txns[i] = &ep.TxnGroup[i].SignedTxn.Txn
+	}
+	policy := resourcePolicy{
+		assignment: makeGroupResourceAssignment(txns, ep.Proto),
```

Comment:
Looks like this the only caller of `makeGroupResourceAssignment`, so you could make it take SignedTxn and save the work of copying the txns.

Reply:
Good point, that's a bit simpler. Done in 031c7c8d2